### ISSUE TITLE
some srcmounts fail to show up in storagepool on boot (mount via fstab)

### ISSUE THREAD

**Issue by muellerflorian01** [trapexit/mergerfs#487]
First of all, thank you so much for putting all the effort in creating mergerfs - it is a really great tool that has solved a long standing issue for me.

General description
===================
I use the following line in fstab to mount three srcmounts:

/mnt/HDD4TB/data:/mnt/HDD2TB/data:/mnt/HDD160GB/data /mnt/datapool fuse.mergerfs defaults,allow_other,direct_io,use_ino,hard_remove,category.create=eplfs,moveonenospc=true,minfreespace=10G,fsname=datapool 0 0

However, if I check the mergerfs mount via "mergerfs.ctl info" after booting, /mnt/HDD2TB/data is not part of the mergerfs mount. The individual devices behind the srcmounts all get mounted correctly, that is lsblk lists ALL the scrmounts drives (/mnt/HDD4TB, /mnt/HDD2TB, /mnt/HDD160GB) as mounted.

If I then unmount /mnt/datapool and  mount it again everything is fine, so syntax or permission problems seem to be ruled out. I suspect that this is because on boot drives are mounted in parallel and thus /mnt/HDD2TB might not be ready when the mergerfs mount is processed.

I was wondering about the recommended way of dealing with this issue.

Expected behavior
=================
mounting mergerfs pool via fstab should mount all listed srcmounts

Actual behavior
===============
mounting mergerfs pool via fstab fails to include some srcmounts, is fine if mounted manually after boot

**Precise** steps to reproduce the behavior
===========================================

> Explicitly list **all** steps to reproduce. Preferably create a minimal example of the problem using standard command line tools. The more variables (apps, settings, etc.) that are involved the more difficult it is to debug. Also, **please** be sure to have read all of the README. It contains a lot of information regarding known system and user issues.

System information
==================
Please provide as much of the following information as possible:

* [ ] mergerfs version: `mergerfs -V`
mergerfs version: 2.24.2
FUSE library version: 2.9.8-mergerfs
fusermount version: 2.9.4
using FUSE kernel interface version 7.19

* [ ] mergerfs settings: `cat /etc/fstab` or the command line arguments
/mnt/HDD4TB/data:/mnt/HDD2TB/data:/mnt/HDD160GB/data /mnt/datapool fuse.mergerfs defaults,allow_other,direct_io,use_ino,hard_remove,category.create=eplfs,moveonenospc=true,minfreespace=10G,fsname=datapool 0 0

* [ ] Linux version: `uname -a`
4.4.0-87-generic #110-Ubuntu SMP Tue Jul 18 12:55:35 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
* [ ] Versions of any additional software being used
* [ ] List of drives, filesystems, & sizes: `df -h`
Dateisystem         Größe Benutzt Verf. Verw% Eingehängt auf
udev                 4,3G       0  4,3G    0% /dev
tmpfs                887M    9,0M  878M    2% /run
/dev/sda1             14G    1,8G   12G   14% /
tmpfs                4,4G       0  4,4G    0% /dev/shm
tmpfs                5,0M       0  5,0M    0% /run/lock
tmpfs                4,4G       0  4,4G    0% /sys/fs/cgroup
/dev/sdd1            147G    131G  9,0G   94% /mnt/HDD160GB
/dev/sdb1            3,6T    2,3T  1,2T   66% /mnt/HDD4TB
/dev/sda3            1,8T    1,6T  111G   94% /mnt/HDD2TB
tmpfs                887M       0  887M    0% /run/user/1000
datapool             5,6T    4,0T  1,3T   76% /mnt/datapool

* [ ] strace of application having problem: `strace -f -o /tmp/app.strace.txt <cmd>` or `strace -f -p <appPID> -o /tmp/app.strace.txt`
na
* [ ] strace of mergerfs while app tried to do it's thing: `strace -f -p <mergerfsPID> -o /tmp/mergerfs.strace.txt`
na

**Comment by muellerflorian01** [comment]
Thanks for the quick reply, that's what I suspected. I'll look into ways to control the mount sequence and post the results.

**Comment by trapexit** [comment]
I could look into a way to make it so it can be told to ignore paths not found at startup. Perhaps with a special flag where if the glob fails to find anything it treats it as a literal? Could simply things though if anything tries to read from the pool before the others are mounted they'd obviously not be available.

**Comment by muellerflorian01** [comment]
If I understand correctly, this would yield the same outcome as it does now - drives not found during boot don't show up in the storagepool. Currently, there no additional issue aside from the drive not being included in the pool, i.e. there are no errors.

I did some searching and the most convenient solutions would be:

a) delayed mounting via rc.local
Set the mergerfs pool mountline in fstab to "noauto", thus preventing mount during boot (and failing to include "slow-to-mount" drives). Then mount the mergerfs pool later by calling mount from rc.local.

b) set the boot mount procedure to sequential mounting
At least on ubuntu linux, the mountall utility (that handles mounting during boot) is set to mount everything in fstab in parallel (i.e. the "fork" option is set), thus creating the possibility that drives are not ready when the mergerfs pool tries to mount. It should be possible to change this option, however I have not yet found how to do so...

**Comment by trapexit** [comment]
No. The reason the drives don't get included is because mergerfs is globbing the paths. The directory doesn't exist at the start so the glob fails to return anything. What I'm suggesting is making it include the path if the glob returns nothing rather than including nothing. It doesn't have to be available for mergerfs to function. It's just that it gets filtered out at start because it doesn't yet exist.

It's not ideal but it is practical.

**Comment by muellerflorian01** [comment]
I see, sorry for the misunderstanding - I agree that this may be a possible solution. However, what happens if a path is truly unavailable (e.g. removed or dead hdd)? If this will cause problems later, than I’d think that delayed mounting of the mergerfs pool (as posted previously) might be the more elegant option.

**Comment by trapexit** [comment]
Nothing would happen except that an invalid path would be in the list. mergerfs already handles missing paths just fine. Effectively they are skipped due to not finding anything.
